### PR TITLE
Add PhpdocTypesOrderFixer

### PR DESCRIPTION
**Pull request by julienfalque**
A new fixer that sorts the types in PHPDoc annotations like `@param`. For now, it only moves `null` to the first position:
```diff
 /**
- * @return string|null
+ * @return null|string
  */
```
An option allows to move it to the last position instead:
```diff
 /**
- * @return null|string
+ * @return string|null
  */
```

### CODE REVIEW

**Comment by localheinz** [comment]
@julienfalque 

Great idea, like it a lot!

**Comment by GrahamCampbell** [comment]
> For now, it only moves null to the first position:

👎

**Comment by julienfalque** [comment]
@GrahamCampbell Can you be more specific?

**Comment by GrahamCampbell** [comment]
I don't like the idea of moving null first. Should be alphabetic, or null last imo.

**Comment by julienfalque** [comment]
I personnaly prefer having `null` always last, hence the option to enable this behavior. The default value is set to match PHP CS Fixer's own style.

**Comment by keradus** [comment]
@GrahamCampbell, it's so great that we built a tool that is not enforcing everyone to use same ruleset ;)

:+1: for rule and configuration, yet it shall also handle cases like `null|int|bool`
later, it could be enhanced to more general rule like `int|bool|null` -> `null|int|bool` -> `null|bool|int`

File: src/Console/Output/ProcessOutput.php [hunk 1]
```diff
@@ -59,14 +59,14 @@
...
-     * @var int|null
+     * @var null|int
      */
     private $symbolsPerLine;
 
     /**
      * @param OutputInterface $output
      * @param EventDispatcher $dispatcher
-     * @param int|null        $nbFiles
+     * @param null|int        $nbFiles
```
Comment (keradus) [inline src/Console/Output/ProcessOutput.php]: please extract self-CS changes into 2.2 line

File: src/DocBlock/Annotation.php [hunk 2]
```diff
@@ -205,10 +205,9 @@ private function getTypesContent()
                 throw new \RuntimeException('This tag does not support types.');
             }
 
-            $tagSplit = preg_split('/\s*\@'.$name.'\s*/', $this->lines[0]->getContent(), 2);
-            $spaceSplit = preg_split('/\s/', $tagSplit[1], 2);
+            preg_match('/\@'.$name.'\s+([^\s|<]+(?:<.*?>)?(?:\|[^\s|<]+(?:<.*?>)?)*)/', $this->lines[0]->getContent(), $matches);
```
Comment (julienfalque) [inline src/DocBlock/Annotation.php]: This is a bugfix, extracted into #2841 which should be merged first.

File: src/Console/Output/ProcessOutput.php (hunk 1 above)
Comment (julienfalque) [inline src/Console/Output/ProcessOutput.php]: Done: #2842.

I tried applying the fixer to 2.3 branch as well: only one fix. Is it worth a PR for it?

Comment (keradus) [inline src/Console/Output/ProcessOutput.php]: nope

File: README.rst [hunk 3]
```diff
@@ -941,6 +941,15 @@ Choose from the list of available rules:
 
   The correct case must be used for standard PHP types in phpdoc.
 
+* **phpdoc_types_order**
+
+  Sorts PHPDoc types.
+
+  Configuration options:
+
+  - ``null_last`` (``bool``): whether ``null`` should be the last type; defaults to
```
Comment (keradus) [inline README.rst]: from option name, i wouldn't say that `false` would cause null to be first

File: README.rst [hunk 4]
```diff
@@ -941,6 +941,15 @@ Choose from the list of available rules:
 
   The correct case must be used for standard PHP types in phpdoc.
 
+* **phpdoc_types_order**
```
Comment (keradus) [inline README.rst]: is it sorting everything or only `null`? if later, name is too generic

Comment (julienfalque) [inline README.rst]: That's on purpose because I want the fixer to be extended in the future to have more advanced sorting options, which is something you seem to agree with:

> later, it could be enhanced to more general rule like int|bool|null -> null|int|bool -> null|bool|int

Extending it may require to deprecate the current option depending on what will be added, but at least we won't have to deprecate the fixer itself in favor of a new one.

File: README.rst (hunk 3 above)
Comment (julienfalque) [inline README.rst]: Indeed.

What about having two options instead, e.g. `first_types` and `last_types` (there are probably better names) that both accept a list of _type categories_? For now, we would only have the `null` category but we could easily extend that list like (e.g. `['objects', 'scalars', 'null']`) without [changing the options in the future](https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/2825#discussion_r121333649).

Comment (keradus) [inline README.rst]: not good idea IMO, as then one would `first_types=null` and `last_types=null` :D
also, I wouldn't go with that big granularity

what about:
- sort_types: bool (sort all types alphabetically)
- null: [opt] `first|last`
?

Comment (julienfalque) [inline README.rst]: Nice start but not customizable enough IMO because on my projects I would like to sort by category of type instead of alphabetically. Another suggestion:

- `sort_algorithm`:
  - `alphabetically` (ignoring leading `\` if any)
  - `by_category` (from "heaviest" to "lightest", i.e.: classes, `object`, `callable`, `resource`, `array<...>`, `...[]`, `array`, scalars, `mixed`, `null`)
  - `none`
- `sort_order`:
  - `ascending`
  - `descending`
- `null_position` (takes precedence over other options):
  - `always_first`
  - `always_last`
  - `default`

Given this discussion, maybe it's better to implement whatever we agree on in this PR directly?

Comment (keradus) [inline README.rst]: sorting alphabetically in descending order is weird... imo not needed

what's the difference between not setting up `null_position` and setting it with `default` value ?

`by_category` algorithm is a bit overkill imo, we should rather ask user to have one (maybe nullable) type instead of mixed types ;)

-------------

> Given this discussion, maybe it's better to implement whatever we agree on in this PR directly?

imo:
- sort_algorithm: 'alpha' (same, short value as in `ordered_imports` rule for consistency)
- nuill_position: 'always_first, always_last'

is enoguh for this PR. later, we could add other algorithms if needed

Comment (julienfalque) [inline README.rst]: > what's the difference between not setting up null_position and setting it with default value ?

Not setting it would be the same as setting it to `always_first`, as it would be the default value of the option.

> imo:
> - sort_algorithm [opt]: 'alpha' (same, short value as in `ordered_imports` rule for consistency)
> - nuill_position [opt]: 'always_first, always_last'
> is enoguh for this PR. later, we could add other algorithms if needed

👍 (but still with `default` as possible value for `null_position`)

Comment (keradus) [inline README.rst]: ok, cool here :+1:

**Comment by keradus** [comment]
rebased.

**Comment by keradus** [comment]
is it still WIP @julienfalque  ?

**Comment by julienfalque** [comment]
Yes, I still have to implement https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/2825#discussion_r121370276.

File: README.rst (hunk 3 above)
Comment (julienfalque) [inline README.rst]: Done.

File: tests/Fixer/Phpdoc/PhpdocTypesOrderFixerTest.php [hunk 5]
```diff
@@ -0,0 +1,409 @@
...
+            [
+                '<?php /** @method \Bar|Foo foo(\Bar|Foo $foo, \Bar|Foo $bar) */',
+                '<?php /** @method Foo|\Bar foo(Foo|\Bar $foo, Foo|\Bar $bar) */',
+            ],
+            [
+                '<?php /** @var array|Bar\Baz|bool[]|false|Foo|int|object|resource|string|string[]|null */',
+                '<?php /** @var string[]|resource|false|object|null|Foo|Bar\Baz|bool[]|string|array|int */',
+            ],
+            [
+                '<?php /** @var array<int, string>|null Foo */',
```
Comment (SpacePossum) [inline tests/Fixer/Phpdoc/PhpdocTypesOrderFixerTest.php]: missing nested cases;
```
array<array<int, array<int, Foo|null>>, array<int,NULL|float>
```

Comment (julienfalque) [inline tests/Fixer/Phpdoc/PhpdocTypesOrderFixerTest.php]: There is one below this one. Did you mean to sort collection key/value types as well, e.g.:
```diff
-/** @var array<Foo|null> */
+/** @var array<null|Foo> */
```

Comment (keradus) [inline tests/Fixer/Phpdoc/PhpdocTypesOrderFixerTest.php]: it would be nice indeed ;)

File: tests/Fixer/Phpdoc/PhpdocTypesOrderFixerTest.php [hunk 6]
```diff
@@ -0,0 +1,409 @@
...
+                '<?php /** @var ?Foo|Bar */',
+            ],
+            [
+                '<?php /** @var ?\Bar|Foo */',
+                '<?php /** @var Foo|?\Bar */',
+            ],
+            [
+                '<?php /** @var Bar|?\Foo */',
+                '<?php /** @var ?\Foo|Bar */',
+            ],
```
Comment (SpacePossum) [inline tests/Fixer/Phpdoc/PhpdocTypesOrderFixerTest.php]: please add tests for:
```
@var a null|\null
@var b null|null
@var c \null|null
@var d \null|\null
```

non-should be fixed, non should mark the Tokens collection as being marked as changed.

Please add a test for
```php
@var a \null|int
@var b array<\null|int>
@var c array<array<array<array<\null|int>\null|int>\null|int>\null|int>>>>
```

File: tests/Fixer/Phpdoc/PhpdocTypesOrderFixerTest.php (hunk 5 above)
Comment (julienfalque) [inline tests/Fixer/Phpdoc/PhpdocTypesOrderFixerTest.php]: Done.

File: tests/Fixer/Phpdoc/PhpdocTypesOrderFixerTest.php (hunk 6 above)
Comment (julienfalque) [inline tests/Fixer/Phpdoc/PhpdocTypesOrderFixerTest.php]: Added all these test cases but:
```php
<?php /** @var array<array<array<array<\null|int>\null|int>\null|int>\null|int>>>> */
```
which looks invalid. Is that on purpose?

Comment (SpacePossum) [inline tests/Fixer/Phpdoc/PhpdocTypesOrderFixerTest.php]:
```php
    /**
     * @var array<int, array<int, array<int, array<int, array<OutputInterface|null>>>>>
     */
```

Comment (julienfalque) [inline tests/Fixer/Phpdoc/PhpdocTypesOrderFixerTest.php]: Done.

File: src/Fixer/Phpdoc/PhpdocTypesOrderFixer.php [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+    {
+        foreach ($tokens as $index => $token) {
+            if (!$token->isGivenKind(T_DOC_COMMENT)) {
+                continue;
+            }
+
+            $doc = new DocBlock($token->getContent());
+            $annotations = $doc->getAnnotationsOfType(Annotation::getTagsWithTypes());
+
+            if (empty($annotations)) {
```
Comment (SpacePossum) [inline src/Fixer/Phpdoc/PhpdocTypesOrderFixer.php]: `!count`

File: src/Fixer/Phpdoc/PhpdocTypesOrderFixer.php [hunk 8]
```diff
@@ -0,0 +1,213 @@
...
+    private function sortTypes(array $types)
+    {
+        foreach ($types as $index => $type) {
+            $types[$index] = preg_replace_callback('/^([^<]+)<(?:(.+?)(,\s*))?(.*)>$/', function (array $matches) {
+                return $matches[1].'<'.$this->sortJoinedTypes($matches[2]).$matches[3].$this->sortJoinedTypes($matches[4]).'>';
+            }, $type);
+        }
+
+        if ('alpha' === $this->configuration['sort_algorithm']) {
+            if (PHP_VERSION_ID < 70000) {
```
Comment (SpacePossum) [inline src/Fixer/Phpdoc/PhpdocTypesOrderFixer.php]: why do you need this?

Comment (julienfalque) [inline src/Fixer/Phpdoc/PhpdocTypesOrderFixer.php]: Because of the way [sorting order of equal elements has changed with PHP 7](http://php.net/manual/en/migration70.incompatible.php#migration70.incompatible.other.sort-order):
> The internal sorting algorithm has been improved, what may result in different sort order of elements, which compare as equal, than before.

On PHP 7, equal elements will remain in the same order while in PHP 5 the order is unpredictable. I'll add a comment with the link above.

File: src/Fixer/Phpdoc/PhpdocTypesOrderFixer.php (hunk 7 above)
Comment (julienfalque) [inline src/Fixer/Phpdoc/PhpdocTypesOrderFixer.php]: Done.

File: src/Fixer/Phpdoc/PhpdocTypesOrderFixer.php (hunk 8 above)
Comment (SpacePossum) [inline src/Fixer/Phpdoc/PhpdocTypesOrderFixer.php]: ah check, they recommend your workaround for PHP7 as well I think;
```
Note:
Don't rely on the order of elements which compare as equal; it might change anytime.
```

Comment (julienfalque) [inline src/Fixer/Phpdoc/PhpdocTypesOrderFixer.php]: I never saw order changes for equal elements on PHP 7 (tests pass) so I assumed current algorithm does maintain correct order and that the note is about future versions of PHP. Shall I apply to "fix" to PHP 7+ anyway?

Comment (SpacePossum) [inline src/Fixer/Phpdoc/PhpdocTypesOrderFixer.php]: I would, as they state it might change anytime, would be a shame to go back and fix this on some other PHP version later on.

Comment (julienfalque) [inline src/Fixer/Phpdoc/PhpdocTypesOrderFixer.php]: Done.

File: src/Fixer/Phpdoc/PhpdocTypesOrderFixer.php [hunk 9]
```diff
@@ -0,0 +1,210 @@
...
+        foreach ($types as $index => $type) {
+            $types[$index] = preg_replace_callback('/^([^<]+)<(?:(.+?)(,\s*))?(.*)>$/', function (array $matches) {
+                return $matches[1].'<'.$this->sortJoinedTypes($matches[2]).$matches[3].$this->sortJoinedTypes($matches[4]).'>';
+            }, $type);
+        }
+
+        if ('alpha' === $this->configuration['sort_algorithm']) {
+            // make sure order of equal elements stays the same
+            foreach ($types as $index => $type) {
+                $types[$index] .= '#'.$index;
```
Comment (keradus) [inline src/Fixer/Phpdoc/PhpdocTypesOrderFixer.php]: implementing stable sorting via `#` is hacky. extract stable sorting to accept any input. as a side note, you are doing Schwartzian transform here. as we do it as well in `FixerFactory::sortFixers`, it's next things that could be extracted.

File: src/Utils.php [hunk 10]
```diff
@@ -126,4 +126,36 @@ public static function calculateTrailingWhitespaceIndent(Token $token)
...
+     *
+     * @param array    $elements
+     * @param callable $getComparedValue a callable that takes a single element and returns the value to compare
+     * @param callable $compareValues    a callable that compares two values
+     *
+     * @return array
+     */
+    public static function stableSort(array $elements, callable $getComparedValue, callable $compareValues)
+    {
+        array_walk($elements, function ($element, $key) {});
```
Comment (keradus) [inline src/Utils.php]: ?

File: src/Utils.php [hunk 11]
```diff
@@ -126,4 +126,35 @@ public static function calculateTrailingWhitespaceIndent(Token $token)
...
+     * Sorts the given values using a Schwartzian transform and a custom
+     * comparison function. Order of equal elements is maintained.
+     *
+     * @param array    $elements
+     * @param callable $getComparedValue a callable that takes a single element and returns the value to compare
+     * @param callable $compareValues    a callable that compares two values
+     *
+     * @return array
+     */
+    public static function stableSort(array $elements, callable $getComparedValue, callable $compareValues)
```
Comment (keradus) [inline src/Utils.php]: input array is modified.

File: src/Utils.php (hunk 10 above)
Comment (julienfalque) [inline src/Utils.php]: Damn, you reviewed faster than I expected :D

File: src/Utils.php (hunk 11 above)
Comment (julienfalque) [inline src/Utils.php]: It's not, but I could change the signature to do that.

Comment (SpacePossum) [inline src/Utils.php]: ? you do modify the input (ok with me though)

File: src/Utils.php [hunk 12]
```diff
@@ -126,4 +126,35 @@ public static function calculateTrailingWhitespaceIndent(Token $token)
...
+        });
+
+        usort($elements, function ($a, $b) use (&$compareValues) {
+            $comparison = $compareValues($a[2], $b[2]);
+
+            if (0 !== $comparison) {
+                return $comparison;
+            }
+
+            return self::cmpInt($a[1], $b[1]);
```
Comment (SpacePossum) [inline src/Utils.php]: for input `$a = ['z' => 1, 'b' => 1, 'c' => 3];` `$a[1]` etc. will not be an int

Comment (julienfalque) [inline src/Utils.php]: Yes, the method only supports integer keys for now. It's internal and we only pass it arrays with integer keys only.

File: README.rst [hunk 13]
```diff
@@ -1002,6 +1002,17 @@ Choose from the list of available rules:
...
+
+  Sorts PHPDoc types.
+
+  Configuration options:
+
+  - ``null_position`` (``'always_first'``, ``'always_last'``, ``'default'``): forces the
+    position of ``null``; defaults to ``'always_first'``
+  - ``sort_algorithm`` (``'alpha'``, ``'none'``): the sorting algorithm to apply;
```
Comment (GrahamCampbell) [inline README.rst]: How about primitives first/last, in order, then objects, in order, either before/after primitives?

**Comment by GrahamCampbell** [comment]
Priority integration test missing. This fixer needs to run after the fixer that changes the case of `null`.

**Comment by GrahamCampbell** [comment]
Are we missing tests that check that only `null`, with that exact casing, is treated as that type?

File: README.rst (hunk 13 above)
Comment (julienfalque) [inline README.rst]: Yeah I wanted to implement something like that, but now I'd rather get this one merged and implement the new sort algorithm in a separate PR as it would probably lead to long discussions :)

**Comment by julienfalque** [comment]
Handling of `null` is case-insensitive so I don't think there is a priority issue here.

**Comment by GrahamCampbell** [comment]
> Handling of null is case-insensitive so I don't think there is a priority issue here.

That is a bug. PHPDoc is treated case sensitive in the rest of php-cs-fixer.

**Comment by GrahamCampbell** [comment]
Apart from in the case of the import fixers checking for references, where we are more conservative.

File: src/Utils.php (hunk 11 above)
Comment (keradus) [inline src/Utils.php]: no, if it is not modifying, it's great

File: src/Utils.php [hunk 14]
```diff
@@ -126,4 +126,36 @@ public static function calculateTrailingWhitespaceIndent(Token $token)
...
+     *
+     * @return mixed[]
+     */
+    public static function stableSort(array $elements, callable $getComparedValue, callable $compareValues)
+    {
+        array_walk($elements, function (&$element, $index) use ($getComparedValue) {
+            $element = [$element, $index, $getComparedValue($element)];
+        });
+
+        usort($elements, function ($a, $b) use (&$compareValues) {
```
Comment (keradus) [inline src/Utils.php]: why reference ?

Comment (julienfalque) [inline src/Utils.php]: Removed.

File: src/Utils.php (hunk 12 above)
Comment (SpacePossum) [inline src/Utils.php]: thanks for the doc updates :) :+1:

File: README.rst [hunk 15]
```diff
@@ -1013,6 +1013,17 @@ Choose from the list of available rules:
 
   The correct case must be used for standard PHP types in phpdoc.
 
+* **phpdoc_types_order**
+
+  Sorts PHPDoc types.
+
+  Configuration options:
+
+  - ``null_position`` (``'always_first'``, ``'always_last'``, ``'default'``): forces the
```
Comment (keradus) [inline README.rst]: what is `null_position => default` ?

File: README.rst [hunk 16]
```diff
@@ -1013,6 +1013,17 @@ Choose from the list of available rules:
...
+    position of ``null``; defaults to ``'always_first'``
+  - ``sort_algorithm`` (``'alpha'``, ``'none'``): the sorting algorithm to apply;
+    defaults to ``'none'``
```
Comment (keradus) [inline README.rst]: hm... what about `defaults to 'alpha'` ?
cc @SpacePossum @juliendufresne

File: README.rst (hunk 15 above)
Comment (julienfalque) [inline README.rst]: `default` keeps the position of `null` as defined by the sort algorithm, unlike `always_first` and `always_last` which override it.

File: README.rst (hunk 16 above)
Comment (julienfalque) [inline README.rst]: I'm not sure most people would use `alpha`, so it wouldn't be a sane default.

I'm not a big fan of alphabetical sorting TBH, I would rather use something like https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/2825#discussion_r124588919.

File: README.rst (hunk 15 above)
Comment (SpacePossum) [inline README.rst]: and what if the sort algorithm is configured `none`, would this be a case of misconfiguration?

File: README.rst (hunk 16 above)
Comment (SpacePossum) [inline README.rst]: I'm :+1: for `alpha`
Btw. how does `alpha` handle pending `\` and casing , like `\int|AbstractTest|\AbstractElseNode|BackUp` ?

File: README.rst (hunk 15 above)
Comment (julienfalque) [inline README.rst]: It's not. We could add validation for that but we do have fixers that don't complain when they are configured to do nothing.

File: README.rst (hunk 16 above)
Comment (julienfalque) [inline README.rst]: Leading `\` and `?` (e.g. `?Foo`) are ignored and sorting in case-insensitive.

So, `alpha` as default? Do we apply it to our codebase as well?

File: README.rst (hunk 15 above)
Comment (SpacePossum) [inline README.rst]: Yeah I know, I'm in fav. of moving away from that. Configuring fixers typically means you want them to do something ;) 
Besides it is also a performance impact as the fixer would still be constructed, configured and tested for candidates, something the fixer is even run as well (like here, the candidate check could see the fixer is configured to do nothing and return `false`)
Throwing the misconfigured exception would be better UX/DX IMHO, however for the current ones it could be a BC break depending on your view of feature vs. bug.

Comment (julienfalque) [inline README.rst]: To me it's a validation that we should always do so lack of it is a bug. I totally agree it would be a UX improvement.

If we decide to add this validation, I think we should do it in a separate PR to apply it everywhere at once.

Comment (keradus) [inline README.rst]: if the fixer is configured to do nothing, then `isCandidate` shall return `false`.
i already started applying this approach.

we could raise a warning on CLI output that some fixer is configured to do nothing, but I am :-1 for raising error.

--------------

@juliendufresne , if `default` means to not move it on top of sorting algorithm, then name is misleading. for me, `default` is meaningful, like `by default we move the null to...`.

File: README.rst (hunk 16 above)
Comment (keradus) [inline README.rst]: yep, I would keep it as that. if one want to have not-standardized order, he could configure the fixer or not use it. idea of fixing is to keep code as similar across whole project of given user as possible. for that, configuring to not update the code is another direction.

File: README.rst (hunk 15 above)
Comment (julienfalque) [inline README.rst]: What would you suggest instead of `default`?

Comment (keradus) [inline README.rst]: what about `null_adjustment=always_first|always_last|none` ?

Comment (julienfalque) [inline README.rst]: Option updated.

File: README.rst (hunk 16 above)
Comment (julienfalque) [inline README.rst]: Defaults updated and applied.

**Comment by julienfalque** [comment]
I added a few test cases that show how the fixer behaves when sorting types that end with `[]` or contain `\` or `_`.

File: src/Utils.php [hunk 17]
```diff
@@ -126,4 +126,36 @@ public static function calculateTrailingWhitespaceIndent(Token $token)
 
         return ltrim($str, "\n");
     }
+
+    /**
+     * Sorts the given values using a Schwartzian transform and a custom
```
Comment (keradus) [inline src/Utils.php]: what about replacing summary with:

```
Perform stable sorting using provided comparison function.

Stability is ensured by using Schwartzian transform.
```

?

Comment (julienfalque) [inline src/Utils.php]: Done.

**Comment by keradus** [comment]
Happy to finally have it ;)

**Comment by keradus** [comment]
Thank you @julienfalque.
